cpp/src/arrow/util/armce-util.h

```diff
@@ -0,0 +1,106 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_UTIL_ARMCE_UTIL_H
+#define ARROW_UTIL_ARMCE_UTIL_H
+
+#include "my_config.h"
+
+namespace arrow {
+
+#if defined(__GNUC__) && defined(__linux__) && defined(ARROW_HAVE_ARMCE)
+
+#include <sys/auxv.h>
+#include <asm/hwcap.h>
+#ifndef HWCAP_CRC32
+#define HWCAP_CRC32 (1 << 7)
+#endif
+static inline uint32_t crc32c_runtime_check(void)
+{
+        unsigned long auxv = getauxval(AT_HWCAP);
+        return (auxv & HWCAP_CRC32) != 0;
+}
+
+#ifdef HAVE_ARMV8_CRC_INTRINSICS
+// compiler intrinsics.
+#include <arm_acle.h>
+#include <arm_neon.h>
+
+#define ARMCE_crc32_u8   __crc32cb
+#define ARMCE_crc32_u16  __crc32ch
+#define ARMCE_crc32_u32  __crc32cw
+#define ARMCE_crc32_u64  __crc32cd
+
+#else
+// Request crc extension capabilities from the assembler
+asm(".arch_extension crc");
+
+// define our own implementations of the intrinsics instead.
+static inline uint32_t ARMCE_crc32_u8(uint32_t crc, uint8_t value) {
+  __asm__("crc32cb %w[c], %w[c], %w[v]":[c]"+r"(crc):[v]"r"(value));
```

Comment:
I'd rather use the official intrinsics if possible. Inline assembler adds a maintenance burden.

Reply:
Inline asm is for the compilers that do not support intrinsics.